port-chooser/src/main/java/org/terracotta/utilities/test/net/PortManager.java

```diff
@@ -0,0 +1,568 @@
+      }
+    }
+  }
+
+  /**
+   * Releases {@code PortRef} instances no longer strongly held.  These {@code PortRef}
+   * instances are considered available for reuse.
+   */
+  private void cleanReleasedPorts() {
+    Reference<? extends PortRef> reference;
+    while ((reference = dereferencedPorts.poll()) != null) {
+      if (reference instanceof AllocatedPort) {
+        AllocatedPort allocatedPort = (AllocatedPort)reference;
+        LOGGER.trace("Port {} dereferenced; releasing reservation", allocatedPort.port);
+        allocatedPort.release();
+      } else {
+        LOGGER.warn("Unexpected Reference observed: {}", reference);
+      }
+    }
+  }
+
+  /**
+   * Reserve and vet the specified candidate port.
+   *
+   * @param candidatePort the port to vet and reserve
+   * @return a new {@code PortRef} instance is the reservation was successful;
+   *      {@code null} otherwise
+   * @throws IllegalStateException if port reservation fails due to a locking error
+   */
+  @SuppressWarnings("try")
+  private PortRef reserveInternal(int candidatePort) {
+    LOGGER.trace("Vetting port {}", candidatePort);
+
+    PortRef portRef = new PortRef(candidatePort);
+    /*
+     * We now have a candidate port; vet the port to ensure it is actually
+     * available.
+     */
+    boolean locked = false;
+    try {
+      // 1) Mark the port reserved in presumption the vetting will succeed
+      portMap.set(candidatePort);
+      allocatedPorts.put(portRef.port, new AllocatedPort(portRef, dereferencedPorts));
+      portRef.onClose(() -> release(candidatePort));
+
+      // 2) Create a server Socket for the candidate port
+      try (ServerSocket ignored = new ServerSocket(candidatePort)) {
+        // 3) While the server socket is open, obtain the system-wide lock for the candidate port
+        locked = systemLevelLocker.lock(portRef);
+      }
+
+      /*
+       * 4) If the port is locked, check the client-side connectability; Windows
+       *    has some odd behaviors in which an outstanding listener isn't required
+       *    for a port to be connectable.
+       */
+      if (locked) {
+        if (refusesConnect(candidatePort)) {
+          LOGGER.debug("Reserved port {}", candidatePort);
+          return portRef;
+        } else {
+          locked = false;   // Cause PortRef close
+          LOGGER.trace("Port {} failed refusesConnect", candidatePort);
+        }
+      } else {
+        LOGGER.trace("Port {} failed to obtain system-level lock", candidatePort);
+      }
+
+    } catch (IllegalStateException e) {
+      // Vetting failed for a non-recoverable reason
+      LOGGER.error("Failed to reserve port {}; abandoning reservation", candidatePort, e);
+      throw e;
+
+    } catch (BindException e) {
+      // Port is in use ...
+
+    } catch (Exception e) {
+      // Some other reservation failure
+      LOGGER.warn("Failed to reserve port {}; checking next port", candidatePort, e);
+
+    } finally {
+      if (!locked) {
```

Comment:
This structure bothers me... any way we can make it a catch-and-rethrow?

Reply:
Added an additional condition and hopefully clarified some names.  Since the method has three exit conditions, catch-and-rethrow doesn't really fit.